File: mvvmcore/src/main/java/com/handpay/mvvm/BaseVMActivity.java
package com.handpay.mvvm;

import android.os.Bundle;

import androidx.annotation.LayoutRes;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProviders;

public abstract class BaseVMActivity<VM extends BaseVM> extends AppCompatActivity {

    public VM vm;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(getContentLayoutResId());
        if (providerVM() != null) {
            vm = ViewModelProviders.of(this).get(providerVM());
        }
        initView(savedInstanceState);
        initData();
    }

    protected abstract Class<VM> providerVM();

    protected abstract void initData();

    protected abstract void initView(@Nullable Bundle savedInstanceState);

    @LayoutRes
    protected abstract int getContentLayoutResId();

}

File: mvvmcore/src/main/java/com/handpay/mvvm/BaseVMActivity.kt
package com.handpay.mvvm

import android.os.Bundle
import androidx.annotation.LayoutRes
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.ViewModelProviders

abstract class BaseVMActivity<VM : BaseVM?> : AppCompatActivity() {
    var model: VM? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(contentLayoutResId)
        providerVM()?.let {
            model=ViewModelProviders.of(this).get(it)
        }
        initView(savedInstanceState)
        initData()
    }

    protected fun providerVM(): Class<VM>? = null
    protected abstract fun initData()
    protected abstract fun initView(savedInstanceState: Bundle?)
    @get:LayoutRes
    protected abstract val contentLayoutResId: Int
}